### PR TITLE
GMG: shared MatrixFree objects

### PR DESCRIPTION
**Pull request by tjhei**
Share the MatrixFree objects between StokesOperator, ABlockOperator, and SchurcomplementOperator for GMG.

This forces batching and iteration of cells in the same order. We assumed beforehand that they would happen in the same order, but this is not always the case! This fixes a problem that in certain situations wrong CellData (viscosity etc.) was being read leading to slow solver convergence (or no convergence).

This also reduces setup cost and memory consumption and requires less code. Win, win, win!?

### CODE REVIEW

**Comment by tjhei** [comment]
possibly related to #4955 , #4563, #4248 and more.

**Comment by MFraters** [comment]
This sound great! I will test out the changes :)

File: source/simulator/stokes_matrix_free.cc [hunk 1]
```diff
@@ -2540,51 +2538,22 @@ namespace aspect
             additional_data.mapping_update_flags = (update_gradients | update_JxW_values |
```
Comment (gassmoeller) [inline source/simulator/stokes_matrix_free.cc:2538]: why this change from value to gradients? is it because we now need gradients and values? should we then specify both?

File: source/simulator/stokes_matrix_free.cc [hunk 2]
```diff
@@ -2444,94 +2446,90 @@ namespace aspect
...
+                // let Dirichlet values win over no normal flux:
+                level_constraints_v.merge(user_level_constraints, AffineConstraints<double>::left_object_wins);
+                level_constraints_v.close();
+              }
+          }
+          {
+            IndexSet relevant_dofs;
+            DoFTools::extract_locally_relevant_level_dofs(dof_handler_p, level, relevant_dofs);
+            level_constraints_p.reinit(relevant_dofs);
+            // TODO: do we need level_constraints_p.add_lines(mg_constrained_dofs_Schur_complement.get_boundary_indices(level));
```
Comment (gassmoeller) [inline source/simulator/stokes_matrix_free.cc:2527]: is this TODO a question? is it necessary for finishing this PR or something for the future?

**Review by gassmoeller — APPROVED** [review]
Looks good. I have two questions, likely just things I didnt understand fully. Feel free to merge if my questions are irrelevant.

File: source/simulator/stokes_matrix_free.cc (hunk 2 above)
Comment (tjhei) [reply source/simulator/stokes_matrix_free.cc:2527]: Good point. It was a question for myself and the answer is: no, not needed.

File: source/simulator/stokes_matrix_free.cc (hunk 1 above)
Comment (tjhei) [reply source/simulator/stokes_matrix_free.cc:2538]: Good catch. Let me clean this up.

Comment (tjhei) [reply source/simulator/stokes_matrix_free.cc]: it turns out that the flags were incorrect anyways (it turns out _values doens't matter).

**Comment by tjhei** [comment]
updated.
